### ISSUE TITLE
Some issues with migrating to one-shot-bufferize

### ISSUE THREAD

**Issue by zjgarvey** [llvm/torch-mlir#3399]
In #3396, I'm running into some issues migrating the [refbackend pipeline](https://github.com/llvm/torch-mlir/blob/a5d3b546f86f9547049b5ee4562bb4698de50b38/projects/pt1/python/torch_mlir_e2e_test/linalg_on_tensors_backends/refbackend.py#L137) to use one-shot-bufferize instead of dialect-specific bufferization passes. 

Relevant section (bufferization) of the original pipeline:

```python
            # MLIR Sparsifier mini-pipeline. Note that this is the bare minimum
            # to ensure operations on sparse tensors are lowered to loops.
            "sparse-assembler{direct-out}",
            "sparsification-and-bufferization",
            "sparse-storage-specifier-to-llvm",
            # Buffer deallocation pass does not know how to handle realloc.
            "func.func(expand-realloc)",
            # Generalize pad and concat after sparse compiler, as they are handled
            # differently when the operations involve sparse operand.
            "func.func(refback-generalize-tensor-pad)",
            "func.func(refback-generalize-tensor-concat)",
            # Bufferize.
            "func.func(scf-bufferize)",
            "func.func(tm-tensor-bufferize)",
            "func.func(empty-tensor-to-alloc-tensor)",
            "func.func(linalg-bufferize)",
            "func-bufferize",
            "arith-bufferize",
            "refback-mlprogram-bufferize",
            "func.func(tensor-bufferize)",
            "func.func(finalizing-bufferize)",
            "func.func(buffer-deallocation)",
```

I've tried many variations, but for example, 

```python
            # MLIR Sparsifier mini-pipeline. Note that this is the bare minimum
            # to ensure operations on sparse tensors are lowered to loops.
            "sparse-assembler{direct-out}",
            "sparsification-and-bufferization",
            "sparse-storage-specifier-to-llvm",
            # Buffer deallocation pass does not know how to handle realloc.
            "func.func(expand-realloc)",
            # Generalize pad and concat after sparse compiler, as they are handled
            # differently when the operations involve sparse operand.
            "func.func(refback-generalize-tensor-pad)",
            "func.func(refback-generalize-tensor-concat)",
            # Bufferize.
            "func.func(tm-tensor-bufferize)",
            "func.func(empty-tensor-to-alloc-tensor)",
            "one-shot-bufferize{dialect-filter=scf,linalg,func,arith,tensor}",
            "refback-mlprogram-bufferize",
            "func.func(finalizing-bufferize)",
            "func.func(buffer-deallocation)",
```

But I keep getting errors with [sparse_test_network](https://github.com/llvm/torch-mlir/blob/a5d3b546f86f9547049b5ee4562bb4698de50b38/test/python/fx_importer/sparse_test.py#L574).

I've tried several variations of the pipeline:

- If not using any dialect-filter, I have issues with `bufferization.to_tensor` during `one-shot-bufferize`, which gets created by the `sparsification-and-bufferization` pass.
- If using a dialect-filter (as above), I have issues with `bufferization.alloc_tensor` at some point in a future pass (in the above example, it fails `finalizing-bufferize`. 
- If using `one-shot-bufferize` before `sparsification-and-bufferization`, I run into issues with sparse ops not wanting to bufferize. 

I'm a bit out of water with all of this. Could someone familiar with sparsity/bufferization help debug this issue?

**Comment by Groverkss** [comment]
Try `one-shot-bufferize{dialect-filter=scf,linalg,func,arith,tensor,bufferization}`

**Comment by zjgarvey** [comment]
> Try `one-shot-bufferize{dialect-filter=scf,linalg,func,arith,tensor,bufferization}`

Thanks for the suggestion, Kunwar. I've tried this, and `bufferize.alloc_tensor` still fails on `finalizing_bufferize`. I've also tried not using any dialect filter, and also using "allow-unknown-ops=1".

**Comment by stellaraccident** [comment]
This is all in llvm. Might want to bring this issue to the author's attention in case if they have an insight. I doubt there are monitoring here.

**Comment by zjgarvey** [comment]
> This is all in llvm. Might want to bring this issue to the author's attention in case if they have an insight. I doubt there are monitoring here.

Okay, I'll post an issue in llvm to try and reach out to some sparsity people.

**Comment by Groverkss** [comment]
Could you also try `--one-shot-bufferize="copy-before-write bufferize-function-boundaries function-boundary-type-conversion=identity-layout-map"` ? This was the command I used to fix upstream sparsification tests. If this also does not work, then this is a bug in sparse pipeline depending on depreciated bufferization passesd.

**Comment by zjgarvey** [comment]
> Could you also try `--one-shot-bufferize="copy-before-write bufferize-function-boundaries function-boundary-type-conversion=identity-layout-map"` ? This was the command I used to fix upstream sparsification tests. If this also does not work, then this is a bug in sparse pipeline depending on depreciated bufferization passes.

Thanks, I haven't tried some of those options yet. I'll let you know.

**Comment by zjgarvey** [comment]
@Groverkss Thanks!!! Those options seem to work. Currently, the pipeline I'm using is: [https://github.com/llvm/torch-mlir/blob/529c88181cfb3b8c3116b10e63ab5c1e20a7e56e/projects/pt1/python/torch_mlir_e2e_test/linalg_on_tensors_backends/refbackend.py#L137-L206](https://github.com/llvm/torch-mlir/blob/529c88181cfb3b8c3116b10e63ab5c1e20a7e56e/projects/pt1/python/torch_mlir_e2e_test/linalg_on_tensors_backends/refbackend.py#L137-L206)

Does anything in the bufferization section of this pipeline seem weird to you? We use this pipeline for nearly all e2e tests, so I wouldn't want to have any unnecessary passes if possible.

**Comment by Groverkss** [comment]
> @Groverkss Thanks!!! Those options seem to work. Currently, the pipeline I'm using is:
> 
> https://github.com/llvm/torch-mlir/blob/529c88181cfb3b8c3116b10e63ab5c1e20a7e56e/projects/pt1/python/torch_mlir_e2e_test/linalg_on_tensors_backends/refbackend.py#L137-L206
> 
> Does anything in the bufferization section of this pipeline seem weird to you? We use this pipeline for nearly all e2e tests, so I wouldn't want to have any unnecessary passes if possible.
> 
> Edit: I needed to move the `one-shot-bufferize` past `tm-tensor-bufferize` and `empty-tensor-to-alloc-tensor`, apparently.

You should remove tm-tensor-bufferize and empty-tensor-to-alloc-tensor also

**Comment by zjgarvey** [comment]
> You should remove tm-tensor-bufferize and empty-tensor-to-alloc-tensor also

I think I need to keep the `tm-tensor-bufferize` (which is apparently a pass registered in torch-mlir). I tried getting rid of it (and also tried using `allow-unknown-ops=1`), but the e2e tests which lower to tm tensor fail on bufferization. I was able to remove the `empty-tensor-to-alloc-tensor` though.

**Comment by Groverkss** [comment]
> > You should remove tm-tensor-bufferize and empty-tensor-to-alloc-tensor also
> 
> I think I need to keep the `tm-tensor-bufferize` (which is apparently a pass registered in torch-mlir). I tried getting rid of it (and also tried using `allow-unknown-ops=1`), but the e2e tests which lower to tm tensor fail on bufferization. I was able to remove the `empty-tensor-to-alloc-tensor` though.

That seems about right then. torch-mlir is still using the depreciated way of doing bufferization. Sad.

**Comment by zjgarvey** [comment]
> > > You should remove tm-tensor-bufferize and empty-tensor-to-alloc-tensor also
> > 
> > 
> > I think I need to keep the `tm-tensor-bufferize` (which is apparently a pass registered in torch-mlir). I tried getting rid of it (and also tried using `allow-unknown-ops=1`), but the e2e tests which lower to tm tensor fail on bufferization. I was able to remove the `empty-tensor-to-alloc-tensor` though.
> 
> That seems about right then. torch-mlir is still using the depreciated way of doing bufferization. Sad.

Yeah :( , `tm-tensor-bufferize` and `refback-mlprogram-bufferize` should both be updated.
